connector/src/main/java/org/geysermc/connector/network/session/cache/PistonCache.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.connector.network.session.cache;
+
+import com.nukkitx.math.vector.Vector3d;
+import com.nukkitx.math.vector.Vector3f;
+import com.nukkitx.math.vector.Vector3i;
+import com.nukkitx.protocol.bedrock.packet.SetEntityMotionPacket;
+import it.unimi.dsi.fastutil.objects.Object2ObjectMaps;
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import lombok.Getter;
+import lombok.Setter;
+import org.geysermc.connector.entity.player.SessionPlayerEntity;
+import org.geysermc.connector.network.session.GeyserSession;
+import org.geysermc.connector.network.translators.collision.BoundingBox;
+import org.geysermc.connector.network.translators.world.block.entity.PistonBlockEntity;
+import org.geysermc.connector.utils.Axis;
+
+import java.util.Map;
+
+public class PistonCache {
+    private final GeyserSession session;
+
+    /**
+     * Maps the position of a piston to its block entity
+     */
+    private final Map<Vector3i, PistonBlockEntity> pistons = Object2ObjectMaps.synchronize(new Object2ObjectOpenHashMap<>());
+
+    /**
+     * Maps the position of a moving block to the piston moving it
+     */
+    @Getter
+    private final Map<Vector3i, PistonBlockEntity> movingBlocksMap = Object2ObjectMaps.synchronize(new Object2ObjectOpenHashMap<>());
+
+    @Getter
+    private Vector3d playerDisplacement = Vector3d.ZERO;
+
+    @Getter @Setter
+    private Vector3f playerMotion = Vector3f.ZERO;
+
+    /**
+     * Stores whether a player has/will collide with any moving blocks.
+     */
+    @Getter @Setter
+    private boolean playerCollided = false;
+
+    /**
+     * Stores whether a player has/will collide with any slime blocks.
+     * This is used to prevent movement from being canceled when players
+     * are about to hit a slime block.
+     */
+    @Getter @Setter
+    private boolean playerSlimeCollision = false;
+
+    /**
+     * Stores whether a player is standing on a honey block.
+     * This is used to cancel movement from Bedrock to prevent them from
+     * falling off.
+     */
+    @Getter @Setter
+    private boolean playerAttachedToHoney = false;
+
+    public PistonCache(GeyserSession session) {
+        this.session = session;
+    }
+
+    public void tick() {
+        resetPlayerMovement();
```

Comment:
Resetting needs to be done at the start because the flags in `PistonCache` is used between ticks in `CollisionManager::adjustBedrockPosition`. I've added an isEmpty check.